doc/designs/evp_skey.md

```diff
@@ -0,0 +1,183 @@
+
+We do not see why we would want to assign an algorithm immediately during
+the creation of an EVP_SKEY as result of derive() operation, if the operation
+just creates bunch of bytes that can be used as a key equally well for HMAC or
+AES encryption.
+
+Objects relationships
+---------------------
+
+We have EVP_CIPHER object associated with a particular provider. A EVP_SKEY
+object is derived for the particular EVP_CIPHER object. After that it is used
+within a EVP_CIPHER_CTX object associated with the same EVP_CIPHER object.
+
+The way to reuse the key within the other provider is through the export/import
+dance. As mentioned before, it may fail.
+
+Key management
+--------------
+
+The provider operating EVP_SKEY objects has to implement a separate key
+management to deal with opaque object. We already have MAC legacy key
+management and it looks suitable for dealing with EVP_SKEY objects.
+
+Allocation and freeing
+--------
+
+```C
+EVP_SKEY * EVP_SKEY_new(void);
+void EVP_SKEY_free(EVP_SKEY *key);
+```
+
+Importing and exporting parameters
+----------------------------------
+
+```C
+int EVP_SKEY_import(EVP_CIPHER *cipher, EVP_SKEY **pskey, OSSL_PARAM params[]);
+int EVP_SKEY_export(const EVP_SKEY *key, OSSL_PARAM **params,
+                    OSSL_CALLBACK *export_cb, void *export_cbarg);
+```
+
+A parameter `raw_bytes` can be used if we try to get a key in classical raw
+bytes implementation.
+
+To find out the parameters supported by a particular EVP_CIPHER object,
+we need a function
+
+```C
+const OSSL_PARAM *EVP_SKEY_fromdata_settable(EVP_CIPHER *cipher);
+```
+
+similar to `EVP_PKEY_fromdata_settable`
+
+API to derive a symmetric key
+-----------------------------
+
+The derived key can be algorithm-specific or algorithm-agnostic. To specify the
+algorithm binding, the params argument can be used.
+
+```C
+int EVP_PKEY_derive_SKEY(EVP_PKEY_CTX *ctx, EVP_SKEY *pskey,
+                         OSSL_PARAM params[]);
+int EVP_KDF_derive_SKEY(EVP_PKEY_CTX *ctx, EVP_SKEY *pskey,
+                        OSSL_PARAM params[]);
+```
+
+similar to `EVP_PKEY_derive/EVP_KDF_derive`
+
+Duplicating the EVP_SKEY object
+-------------------------------
+
+**Open question:** do we need it?
+
+```C
+EVP_SKEY * EVP_SKEY_dup(const EVP_SKEY * skey);
+```
+
+Using in cipher operations
+--------------------------
+
+We already have the function
+
+```C
+EVP_CipherInit_ex2(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
+                   const unsigned char *key, const unsigned char *iv,
+                   int enc, const OSSL_PARAM params[]);
+```
+
+To set the skey object to the context, we invoke `EVP_CipherInit_ex2` with NULL
+key ptr and invoke `EVP_CIPHER_CTX_set1_EVP_SKEY`.
+
+An alternate approach will be providing a function
+
+```C
+EVP_CipherInit_skey(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
+                   const EVP_SKEY *key,
+                   const unsigned char *iv, size_t iv_len,
```

Comment:
Can we drop the IV as well, not all ciphers have "IVs" or let them be passed in, it would be better to pass IVs via params/ctx_set.